activerecord/lib/active_record/destroy_association_async_job.rb

```diff
@@ -23,8 +23,15 @@ def perform(
         raise DestroyAssociationAsyncError, "owner record not destroyed"
       end
 
-      association_model.where(association_primary_key_column => association_ids).find_each do |r|
-        r.destroy
+      if association_primary_key_column.is_a?(Array)
```

Comment:
We have the association class so we can ask `association_class.query_constraints_list` instead of `is_a?` array. A model will only have query constraints if explicitly defined.